### PR TITLE
Online Hard Node Mining: error-weighted surface loss

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Surface MAE is dominated by a minority of hard-to-predict nodes (suction peaks, stagnation points, TE pressure recovery). Currently, all surface nodes contribute equally to the loss. By weighting each node's loss contribution by its current prediction error (detached), we focus gradient signal on the nodes that matter most for MAE reduction. This is analogous to focal loss for detection (Lin et al. 2017) adapted for regression, and online hard example mining (OHEM, Shrivastava et al. 2016) applied at the node level.

**Key insight:** The nodes with the largest absolute errors dominate the MAE metric. Training the model to prioritize these nodes should reduce MAE more efficiently than uniform weighting.

### 1. Add CLI flags to `Config` dataclass (near other loss flags around line 1100):
```python
ohnm: bool = False                  # online hard node mining for surface loss
ohnm_gamma: float = 1.0             # error exponent (1.0=linear, 2.0=quadratic focus)
```

### 2. Modify the surface loss computation
Find the surface MAE loss calculation in the training loop (where per-node `|pred - target|` is computed for surface nodes). Modify it as follows:

```python
if cfg.ohnm:
    # Compute per-node errors (detached — no gradient through weights)
    with torch.no_grad():
        node_errors = (pred_surface - gt_surface).abs().mean(dim=-1)  # [M] per-node error across channels
        weights = node_errors ** cfg.ohnm_gamma  # [M] focus on hard nodes
        weights = weights / (weights.mean() + 1e-8)  # normalize so mean weight = 1
    # Apply weights to the loss (weights are detached, loss still gets gradients)
    weighted_surface_loss = (weights.unsqueeze(-1) * (pred_surface - gt_surface).abs()).mean()
else:
    weighted_surface_loss = (pred_surface - gt_surface).abs().mean()  # original uniform MAE
```

**Important details:**
- `weights` must be **detached** (no gradient) — they guide the loss but don't participate in backprop
- Only apply OHNM to the **surface MAE loss**, NOT to DCT freq loss, volume loss, or SRF loss
- Apply OHNM to **both** the main surface loss and the aft-foil surface loss (if `--aft_foil_srf` is active)
- The normalization `weights / weights.mean()` ensures the total loss magnitude stays comparable to baseline
- gamma=1.0 means a node with 2x error gets 2x weight; gamma=2.0 means 4x weight
- Apply this BEFORE any per-sample-std normalization to avoid interfering with the existing normalization pipeline

### 3. Run configuration
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/ohnm-gamma1" \
  --wandb_group "ohnm" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --ohnm --ohnm_gamma 1.0 \
  --seed 42
```
Run with seeds 42 and 73.

## Baseline (PR #2213, 2-seed avg)
| Metric | Value | Target to beat |
|--------|-------|---------------|
| p_in | 11.979 | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| p_tan | 28.341 | < 28.34 |
| p_re | 6.300 | < 6.30 |

Baseline W&B runs: hgml7i2r (s42), qic03vrg (s73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### OHNM gamma=1.0, 2-seed

| Metric | Baseline (2-seed avg) | OHNM (seed 42) | OHNM (seed 73) | OHNM (2-seed avg) | Δ vs baseline |
|--------|----------------------|----------------|----------------|-------------------|---------------|
| **p_in** | **11.979** | 11.903 | 12.319 | **12.111** | **+1.1%** ❌ |
| **p_oodc** | **7.643** | 8.414 | 7.988 | **8.201** | **+7.3%** ❌ |
| **p_tan** | **28.341** | 28.755 | 28.816 | **28.786** | **+1.6%** ❌ |
| **p_re** | **6.300** | 6.766 | 6.772 | **6.769** | **+7.4%** ❌ |

### Run details
- **Seed 42**: W&B run `wugogei0`, best epoch 147/500, peak VRAM 46.6GB, 180 min
- **Seed 73**: W&B run `9xl8ixv0`, best epoch 147/500, peak VRAM 46.4GB, 181 min
- Both ran ~147 epochs (EMA from epoch 140)

### Command used
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/ohnm-gamma1" \
  --wandb_group "ohnm" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --ohnm --ohnm_gamma 1.0 \
  --seed {42,73}
```

### What happened

**Negative result.** OHNM with gamma=1.0 degraded all four pressure metrics, most significantly p_oodc (+7.3%) and p_re (+7.4%).

**Why it likely failed:**

1. **Double-stacking with existing hard-node mining:** The codebase already has an asymmetric hard-node mining mechanism (epoch ≥ 30) that gives 1.5× weight to above-median-error surface nodes for non-tandem samples. OHNM stacks multiplicatively on top of this — a hard node could get `(err/mean)^1 × 1.5 = 3-5×` effective weight, over-focusing gradient signal on a small subset of nodes.

2. **Over-focusing hurts generalization:** The OOD tracks (p_oodc, p_re) degraded most, suggesting the model overfits to the specific hard nodes in the training distribution (particular suction peaks, stagnation patterns) at the expense of generalization to unseen conditions.

3. **The baseline is already well-optimized for hard nodes:** With the existing asymmetric mining, surface refinement heads, and pressure-first architecture, the baseline already has strong inductive bias toward hard-to-predict regions. Additional explicit error weighting may be redundant or harmful.

### Suggested follow-ups
- Try OHNM **without** the existing asymmetric hard-node mining (disable the epoch≥30 median-threshold block) to isolate the pure OHNM effect
- Try a **top-k percentile** variant instead of continuous weighting (e.g., weight top 25% hardest nodes at 2× and ignore gamma entirely)
- Try a lower gamma (0.5) to soften the focus and reduce over-weighting

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2249 because OHNM with gamma=1.0 is a clear regression across all four metrics: p_in +1.1%, p_oodc +7.3%, p_tan +1.6%, p_re +7.4% — significantly worse than baseline. W&B runs wugogei0 (s42) and 9xl8ixv0 (s73) confirmed and results match student report exactly.

The student's analysis is well-reasoned and likely correct: the baseline already has asymmetric hard-node mining (1.5× for above-median-error nodes at epoch ≥30), so stacking OHNM multiplicatively over-concentrates gradients on a small subset of nodes, hurting OOD generalization (p_oodc and p_re hurt most — both +7%). The baseline's existing hard-node mining already handles this use case. Closing as a dead end.

Good analysis frieren — the 'already optimized' diagnosis is exactly right. The suggested follow-up of disabling the existing asymmetric mining before trying OHNM is scientifically sound but adds complexity for uncertain gain. Will assign a fresh direction instead..
